src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Variables.cs

```diff
@@ -0,0 +1,434 @@
+            /// tracked variable. Slots less than 0 are illegal.
+            /// </summary>
+            private readonly ArrayBuilder<VariableIdentifier> _variableBySlot = ArrayBuilder<VariableIdentifier>.GetInstance(1, default);
+
+            internal static Variables Create(Symbol? symbol)
+            {
+                return new Variables(id: 0, container: null, symbol);
+            }
+
+            internal static Variables Create(VariablesSnapshot snapshot)
+            {
+                var container = snapshot.Container is null ? null : Create(snapshot.Container);
+                var variables = new Variables(snapshot.Id, container, snapshot.Symbol);
+                variables.Populate(snapshot);
+                return variables;
+            }
+
+            private int GetNextId()
+            {
+                return Id +
+#if DEBUG
+                    _nextIdOffset.Next(maxValue: 7) +
+#endif
+                    1;
+            }
+
+            private void Populate(VariablesSnapshot snapshot)
+            {
+                Debug.Assert(_variableSlot.Count == 0);
+                Debug.Assert(_variableTypes.Count == 0);
+                Debug.Assert(_variableBySlot.Count == 1);
+
+                _variableBySlot.AddMany(default, snapshot.VariableSlot.Length);
+                foreach (var pair in snapshot.VariableSlot)
+                {
+                    var identifier = pair.Key;
+                    var index = pair.Value;
+                    _variableSlot.Add(identifier, index);
+                    _variableBySlot[index] = identifier;
+                }
+
+                foreach (var pair in snapshot.VariableTypes)
+                {
+                    _variableTypes.Add(pair.Key, pair.Value);
+                }
+            }
+
+            private Variables(int id, Variables? container, Symbol? symbol)
+            {
+                Debug.Assert(id >= 0);
+                Debug.Assert(id <= IdMask);
+                Debug.Assert(container is null || container.Id < id);
+#if DEBUG
+                _nextIdOffset = container?._nextIdOffset ?? new Random();
+#endif
+                Id = id;
+                Container = container;
+                Symbol = symbol;
+            }
+
+            internal void Free()
+            {
+                Container?.Free();
+                _variableBySlot.Free();
+                _variableTypes.Free();
+                _variableSlot.Free();
+            }
+
+            internal VariablesSnapshot CreateSnapshot()
+            {
+                return new VariablesSnapshot(
+                    Id,
+                    Container?.CreateSnapshot(),
+                    Symbol,
+                    ImmutableArray.CreateRange(_variableSlot),
+                    ImmutableDictionary.CreateRange(_variableTypes));
+            }
+
+            internal Variables CreateNestedFunction(MethodSymbol method)
```

Comment:
>CreateNestedFunction [](start = 31, length = 20)

The name "CreateNestedFunction" feels confusing. The method doesn't create a function. Consider using the name that reflects the purpose of the method. Given the terminology used in doc comments, "CreateNestedMethodScope" comes to mind.